File: src/main/java/com/mattchowning/file_read_write/client/calls/GetOAuthCall.java
package com.mattchowning.file_read_write.client.calls;

import io.netty.channel.ChannelOutboundInvoker;
import io.netty.handler.codec.http.*;
import io.netty.handler.codec.http.multipart.HttpPostRequestEncoder;

import static com.mattchowning.file_read_write.SharedConstants.*;

public class GetOAuthCall extends OAuthCall {

    private final String username;
    private final String password;

    public GetOAuthCall(String username, String password) {
        super();
        this.username = username;
        this.password = password;
    }

    @Override
    protected void makeRequest(ChannelOutboundInvoker ctx) {
        System.out.println("Requesting OAuth token...");
        try {
            FullHttpRequest request = getOAuthTokenRequest();
            ctx.writeAndFlush(request);
        } catch (HttpPostRequestEncoder.ErrorDataEncoderException e) {
            e.printStackTrace();
            ctx.close();
        }
    }

    private FullHttpRequest getOAuthTokenRequest() throws HttpPostRequestEncoder.ErrorDataEncoderException {
        FullHttpRequest request =  new DefaultFullHttpRequest(HttpVersion.HTTP_1_1,
                                                              HttpMethod.POST,
                                                              OAUTH_PATH);

        HttpPostRequestEncoder postRequestEncoder = new HttpPostRequestEncoder(request, false);
        postRequestEncoder.addBodyAttribute(GRANT_TYPE_KEY, GRANT_TYPE_PASSWORD);
        postRequestEncoder.addBodyAttribute(USERNAME_KEY, username);
        postRequestEncoder.addBodyAttribute(PASSWORD_KEY, password);
        postRequestEncoder.finalizeRequest();
        return request;
    }
}

File: src/main/scala/com/mattchowning_scala/file_read_write/client/calls/GetOAuthCall.scala
package com.mattchowning_scala.file_read_write.client.calls

import com.mattchowning.file_read_write.server.model.OAuthToken
import io.netty.channel.ChannelOutboundInvoker
import io.netty.handler.codec.http._
import io.netty.handler.codec.http.multipart.HttpPostRequestEncoder
import com.mattchowning.file_read_write.SharedConstants._

class GetOAuthCall(username: String,
                   password: String,
                   successCallback: OAuthToken => Unit,
                   failureCallback: () => Unit)
  extends OAuthCall(successCallback, failureCallback) {

  protected def makeRequest(ctx: ChannelOutboundInvoker) {
    System.out.println("Requesting OAuth token...")
    try {
      val request: FullHttpRequest = getOAuthTokenRequest
      ctx.writeAndFlush(request)
    }
    catch {
      case e: HttpPostRequestEncoder.ErrorDataEncoderException =>
        e.printStackTrace()
        ctx.close
    }
  }

  @throws(classOf[HttpPostRequestEncoder.ErrorDataEncoderException])
  private def getOAuthTokenRequest: FullHttpRequest = {
    val request: FullHttpRequest = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.POST, OAUTH_PATH)
    val postRequestEncoder: HttpPostRequestEncoder = new HttpPostRequestEncoder(request, false)
    postRequestEncoder.addBodyAttribute(GRANT_TYPE_KEY, GRANT_TYPE_PASSWORD)
    postRequestEncoder.addBodyAttribute(USERNAME_KEY, username)
    postRequestEncoder.addBodyAttribute(PASSWORD_KEY, password)
    postRequestEncoder.finalizeRequest
    request
  }
}